dali/operators/generic/reduce/histogram.h

```diff
@@ -0,0 +1,248 @@
+// Copyright (c) 2021, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_OPERATORS_GENERIC_REDUCE_HISTOGRAM_H_
+#define DALI_OPERATORS_GENERIC_REDUCE_HISTOGRAM_H_
+
+#include "dali/kernels/kernel_manager.h"
+#include "dali/operators/generic/reduce/axes_helper.h"
+#include "dali/pipeline/operator/operator.h"
+
+namespace dali {
+namespace hist_detail {
+
+inline bool is_simple_reduction(const std::vector<int> &reduction_axes, int ndims,
+                                int hist_dim = 1) {
+  int dim = ndims - hist_dim;
+
+  // Starting from inner dimension, look if we should reduce this axis
+  // If we can do so until we can collapse next dimention.
+  auto raxis = reduction_axes.rbegin();
+
+  for (; raxis != reduction_axes.rend(); ++raxis) {
+    // TODO: Handle multidimentional histograms correctly
+    if (dim != *raxis) {
+      break;
+    }
+    --dim;
+  }
+
+  // If there's no reduction axes left, we won't need transpose any axis.
+  if (raxis == reduction_axes.rend())
+    return true;
+
+  return false;
+}
+
+// Collapses all inner dimensions
```

Comment:
Can't we also collapse all outer dimensions?
After rearranging the axes so that we have
non-reduced   reduced   [channel]
We can then collapse (or expand!) each of these groups, so that we have exactly one non-reduced dimension, one reduced dimension and one channel dimension.
If it happens, that one of these groups is empty, we can always insert a phony axis with extent 1.